The attached patch gives the user ability to define how many symbolic
links rsync should follow before actually adding it to the file list.

The patch has been heavily modified from its original form to work
with the latest codebase, but even in its original form it didn't
handle relative symlinks properly, and that has not yet been fixed
in this modified version.

--- orig/flist.c	2005-07-28 18:48:38
+++ flist.c	2004-07-16 16:58:04
@@ -45,6 +45,7 @@ extern int filesfrom_fd;
 extern int one_file_system;
 extern int keep_dirlinks;
 extern int preserve_links;
+extern int follow_links_depth;
 extern int preserve_hard_links;
 extern int preserve_perms;
 extern int preserve_devices;
@@ -734,6 +735,30 @@ static struct file_struct *receive_file_
 }
 
 
+#if SUPPORT_LINKS
+static int links_depth(char *linkname, STRUCT_STAT *st_ptr)
+{
+	char buf[MAXPATHLEN];
+	STRUCT_STAT st;
+	int i;
+
+	for (i = 0; i < follow_links_depth; i++) {
+		/* XXX This doesn't handle relative symlinks! */
+		if (readlink_stat(linkname, &st, buf) != 0)
+			break;
+		*st_ptr = st;
+		if (!S_ISLNK(st.st_mode))
+			return 1;
+		strlcpy(linkname, buf, MAXPATHLEN);
+#if 0
+		fprintf(stderr, "\n%s:%i [#%i] %s -> %s\n", __FILE__, __LINE__, i, file->u.link, linkname);
+#endif
+	}
+
+	return 0;
+}
+#endif
+
 /**
  * Create a file_struct for a named file by reading its stat()
  * information and performing extensive checks against global
@@ -859,7 +884,13 @@ skip_filters:
 	basename_len = strlen(basename) + 1; /* count the '\0' */
 
 #ifdef SUPPORT_LINKS
-	linkname_len = S_ISLNK(st.st_mode) ? strlen(linkname) + 1 : 0;
+	if (S_ISLNK(st.st_mode)) {
+		if (follow_links_depth && links_depth(linkname, &st))
+			linkname_len = 0;
+		else
+			linkname_len = strlen(linkname) + 1;
+	} else
+		linkname_len = 0;
 #else
 	linkname_len = 0;
 #endif
--- orig/options.c	2005-07-28 18:48:38
+++ options.c	2005-01-28 19:33:24
@@ -42,6 +42,7 @@ int archive_mode = 0;
 int keep_dirlinks = 0;
 int copy_links = 0;
 int preserve_links = 0;
+int follow_links_depth = 0;
 int preserve_hard_links = 0;
 int preserve_perms = 0;
 int preserve_devices = 0;
@@ -276,6 +277,7 @@ void usage(enum logcode F)
   rprintf(F,"     --inplace               update destination files in-place (SEE MAN PAGE)\n");
   rprintf(F," -d, --dirs                  transfer directories without recursing\n");
   rprintf(F," -l, --links                 copy symlinks as symlinks\n");
+  rprintf(F,"     --links-depth=NUM       follow symlinks up to NUM depth\n");
   rprintf(F," -L, --copy-links            transform symlink into referent file/dir\n");
   rprintf(F,"     --copy-unsafe-links     only \"unsafe\" symlinks are transformed\n");
   rprintf(F,"     --safe-links            ignore symlinks that point outside the source tree\n");
@@ -402,6 +404,7 @@ static struct poptOption long_options[] 
   {"inplace",          0,  POPT_ARG_NONE,   &inplace, 0, 0, 0 },
   {"dirs",            'd', POPT_ARG_VAL,    &xfer_dirs, 2, 0, 0 },
   {"links",           'l', POPT_ARG_NONE,   &preserve_links, 0, 0, 0 },
+  {"links-depth",      0,  POPT_ARG_INT,    &follow_links_depth , 0, 0, 0 },
   {"copy-links",      'L', POPT_ARG_NONE,   &copy_links, 0, 0, 0 },
   {"keep-dirlinks",   'K', POPT_ARG_NONE,   &keep_dirlinks, 0, 0, 0 },
   {"whole-file",      'W', POPT_ARG_VAL,    &whole_file, 1, 0, 0 },
--- orig/rsync.yo	2005-07-28 19:26:48
+++ rsync.yo	2005-01-28 19:33:40
@@ -309,6 +309,7 @@ to the detailed description below for a 
      --inplace               update destination files in-place
  -d, --dirs                  transfer directories without recursing
  -l, --links                 copy symlinks as symlinks
+     --links-depth=NUM       follow symlinks up to NUM depth
  -L, --copy-links            transform symlink into referent file/dir
      --copy-unsafe-links     only "unsafe" symlinks are transformed
      --safe-links            ignore symlinks that point outside the tree
